src/generators/notebookgenerator/itemrenderer.tsx

```diff
@@ -13,8 +15,161 @@ import { sanitizerOptions } from '../shared';
 
 import * as React from 'react';
 
+/**
+ * Returns the header level for a provided cell.
+ *
+ * ## Notes
+ *
+ * -   If a cell does not contain a header, the function returns the sentinel value `-1`.
+ *
+ * -   Cell header examples:
+ *
+ *     -   Markdown header:
+ *
+ *         ```
+ *         # Foo
+ *         ```
+ *
+ *     -   Markdown header:
+ *
+ *         ```
+ *         Foo
+ *         ===
+ *         ```
+ *
+ *         ```
+ *         Foo
+ *         ---
+ *         ```
+ *
+ *     -   HTML heading:
+ *
+ *         ```
+ *         <h3>Foo</h3>
+ *         ```
+ *
+ * @private
+ * @param cell - notebook cell
+ * @returns header level
+ */
+function headerLevel(cell: Cell): number {
+  if (cell.constructor.name !== 'MarkdownCell') {
+    return -1;
+  }
+  const lines = cell.model.value.text.split('\n');
+
+  // Case: Markdown header
+  let match = lines[0].match(/^([#]{1,6}) (.*)/);
+  if (match) {
+    return match[1].length;
+  }
+  // Case: Markdown header
+  if (lines.length > 1) {
+    match = lines[1].match(/^([=]{2,}|[-]{2,})/);
+    if (match) {
+      return match[1][0] === '=' ? 1 : 2;
+    }
+  }
+  // Case: HTML heading (WARNING: this is not particularly robust, as HTML headings can span multiple lines)
+  match = lines[0].match(/<h([1-6])>(.*)<\/h\1>/i);
+  if (match) {
+    return parseInt(match[1], 10);
+  }
+  return -1;
+}
+
+/**
+ * Collapses a notebook cell by hiding its section-defined sub-cells (i.e., cells which have lower precedence).
+ *
+ * @private
+ * @param tracker - notebook tracker
+ * @param cell - notebook cell
+ */
+function collapseCell(tracker: any, cell: Cell): void {
```

Comment:
Why is `tracker` `any`? Isn't it of type `INotebookTracker`?

Reply:
I think it may be better to pass in the `Notebook` widget, rather than the `INotebookTracker` here. This makes the universe that this function needs to know about smaller.

Reply:
Addressed.